Sources/SourceKit/sourcekitd/SwiftLanguageServer.swift

```diff
@@ -547,6 +548,81 @@ extension SwiftLanguageServer {
     // FIXME: cancellation
     _ = handle
   }
+
+  func foldingRange(_ req: Request<FoldingRangeRequest>) {
+
+    guard let snapshot = documentManager.latestSnapshot(req.params.textDocument.url) else {
+      log("failed to find snapshot for url \(req.params.textDocument.url)")
+      req.reply(nil)
+      return
+    }
+
+    let skreq = SKRequestDictionary(sourcekitd: sourcekitd)
+    skreq[keys.request] = requests.editor_open
+    skreq[keys.name] = "FoldingRanges:" + snapshot.document.url.path
+    skreq[keys.sourcetext] = snapshot.text
+    skreq[keys.syntactic_only] = 1
+
+    let handle = sourcekitd.send(skreq) { [weak self] result in
+      guard let self = self else { return }
+      guard let dict = result.success else {
+        req.reply(.failure(result.failure!))
+        return
+      }
+
+      guard let syntaxMap: SKResponseArray = dict[self.keys.syntaxmap],
+            let substructure: SKResponseArray = dict[self.keys.substructure] else {
+        return req.reply([])
+      }
+
+      var ranges: [FoldingRange] = []
+
+      syntaxMap.forEach { _, value in
+        if let kind: sourcekitd_uid_t = value[self.keys.kind],
+           kind == self.values.syntaxtype_comment,
+           let offset: Int = value[self.keys.offset],
+           let start: Position = snapshot.positionOf(utf8Offset: offset),
+           let length: Int = value[self.keys.length],
+           // SourceKit marks the end of a comment as the first non-comment character
+           // after it, so we subtract one to get the real comment range.
+           let end: Position = snapshot.positionOf(utf8Offset: offset + length - 1) {
```

Comment:
I don't think this subtraction should be necessary.  Both LSP and sourcekitd use half-open ranges (the range includes the start position and goes up to but *not* including the end position).  They should both expect the end offset to be one-past the last character of the comment AFAICT.

Heres what the spec says about the `Range` type:
> A range in a text document expressed as (zero-based) start and end positions. A range is comparable to a selection in an editor. Therefore the end position is exclusive.

And the comments in the spec about FoldingRange's endCharacter says that it defaults to the line length, which is also one-past the end.

Reply:
The reason I added that is because a simple comment like `//foo` will end up with a different endLine, as the "end" of the comment here is the start of the next line. It was causing VSCode to fold whatever was on the next line even though the comment only really covers a single line. It might have to do with the supposed ignored properties I mentioned earlier in this case, but I wasn't sure how to stop it other than subtracting it. What do you think?

Reply:
Hmm, in that case subtracting 1 _before_ finding the Position doesn't seem like a good fix, since it only does the right thing if the end position is on a different line.  Maybe you could compute the correct Position, but then check if the end of the range is at column 0, and in that case change it to the previous line but with column number = line length?  It's the same position in an abstract sense, but if something is only looking at the line number and not the column it will be more correct.

Reply:
To clarify, I mean change `Position(line: N, utf16Index: 0)` to `Position(line: N-1, utf16Index: <length of line N-1>)`.  You can get the line length from the snapshot's LineTable.

Reply:
Definitely sounds better! To be fair, I just noticed the same problem exists if you have something right after a brace, like a promise:
```
foo().then {
   //
}.then {
}
```
Folding the first block would also fold the second `then` if the client only supports folding whole lines. It's the same problem, although the comment one definitely looks weirder. If you're right about the end position being expected to be one character past then perhaps we don't need to treat this at all since the problem will always exist for bodies.